File: cpp/gql/src/libgql/src/libgql/parsers/schema/nodes/interface.cpp
#include "./interface.hpp"

#include <map>
#include <memory>
#include <ranges>
#include <string>
#include <utility>

#include "../../file/server/ast.hpp"
#include "../nodes/object_field_spec.hpp"
#include "../server_ast.hpp"
#include "../shared_ast.hpp"
#include "../type_registry.hpp"

using namespace gql::parsers::file;

namespace gql::parsers::schema::nodes {
std::shared_ptr<ast::Interface> parseInterface(
    const server::ast::InterfaceDefinitionNode &node,
    const TypeRegistry &registry) {
    const auto &obj = registry.interfaces.at(node.name.name);
    obj->fields =
        node.fields |
        std::views::transform(
            [&registry](const server::ast::FieldDefinitionNode &defNode)
                -> std::pair<std::string, std::shared_ptr<ast::FieldDefinition<
                                              ast::ObjectFieldSpec>>> {
                const auto &[typeSpec, nullable] =
                    parseObjectFieldSpec(defNode, registry);
                return { defNode.name.name,
                         std::make_shared<
                             ast::FieldDefinition<ast::ObjectFieldSpec>>(
                             defNode.name.name, typeSpec, nullable) };
            }) |
        std::ranges::to<std::map>();
    return obj;
};
};  // namespace gql::parsers::schema::nodes

File: rust/libgql/src/parsers/schema/server/interface.rs
use crate::parsers::{
    file,
    schema::{
        server::{directive, errors, object},
        shared,
    },
};

use super::type_registry::HashMapTypeRegistry;

pub fn parse_definition<'buffer>(
    node: &file::server::ast::InterfaceDefinitionNode<'buffer>,
    registry: &mut HashMapTypeRegistry,
) -> Result<(), errors::Error<'buffer>> {
    let mut intermediate = Vec::new();
    for field_definition_node in node.fields.iter() {
        intermediate.push(object::parse_object_field_spec(
            &field_definition_node,
            registry,
        )?);
    }
    let directives = directive::parse_invocations(&node.directives, registry)?;
    let obj = registry.interfaces.get_mut(node.name.name).unwrap();
    for (field_definition_node, (spec, nullable)) in
        node.fields.iter().zip(intermediate)
    {
        obj.fields.insert(
            field_definition_node.name.name.to_string(),
            shared::ast::FieldDefinition {
                name: field_definition_node.name.name.to_string(),
                spec,
                nullable,
            },
        );
    }
    obj.directives = directives;
    return Ok(());
}
